src/server.rs

```diff
@@ -1,21 +1,85 @@
 //! Faucet server implementation.
 
+use std::collections::HashSet;
+use std::net::{IpAddr, ToSocketAddrs as _};
+use std::str::FromStr as _;
+use std::sync::{Arc, RwLock};
+use std::time::{Duration, SystemTime};
+
 use actix_cors::Cors;
 use actix_web::http::{header, StatusCode};
-use actix_web::web::{get, post, Bytes};
-use actix_web::{App, HttpResponse, HttpServer, Responder};
-use eyre::Result;
+use actix_web::web::{get, post, Bytes, Data};
+use actix_web::{App, HttpRequest, HttpResponse, HttpServer, Responder};
+use eyre::{eyre, Result};
 use tracing::{error, info};
 
-use crate::{active_requests, config, erc20_tokens, id, neon_token};
+use crate::{active_requests, config, erc20_tokens, id, neon_token, solana};
+
+type AirdropLimiter = Data<RwLock<neon_token::AirdropLimiter>>;
 
 /// Starts the server in listening mode.
 pub async fn start(workers: usize) -> Result<()> {
     let rpc_bind = config::rpc_bind();
     let rpc_port = config::rpc_port();
     info!("{} Bind {}:{}", id::default(), rpc_bind, rpc_port);
 
-    HttpServer::new(|| {
+    let blacklist = config::blacklisted_ips()
+        .into_iter()
+        .map(|ip| IpAddr::from_str(&ip).map_err(|err| eyre!("Invalid blacklisted ip: {}", err)))
+        .collect::<Result<_>>()?;
+    let mut trusted_proxies = HashSet::new();
+    for uri in config::allowed_origins().into_iter() {
+        let uri = actix_web::http::Uri::from_str(&uri)
+            .map_err(|err| eyre!("Invalid trusted proxy '{}': {}", uri, err))?;
+        let host = uri
+            .host()
+            .ok_or_else(|| eyre!("Invalid trusted proxy '{}': no host", uri))?;
+        let ip = (host, 0)
+            .to_socket_addrs()
+            .map_err(|err| eyre!("Invalid trusted proxy '{}': {}", host, err))?
+            .next()
+            .ok_or_else(|| eyre!("Invalid trusted proxy '{}': lookup failed", host))?
+            .ip();
+        trusted_proxies.insert(ip);
+    }
+    let per_request_cap = solana::convert_whole_to_fractions(config::solana_max_amount())
+        .map_err(|err| eyre!("invalid max amount: {}", err))?;
+    let per_time_cap = solana::convert_whole_to_fractions(config::solana_per_time_max_amount())
+        .map_err(|err| eyre!("invalid per time max amount: {}", err))?;
+    let day_in_secs = config::solana_slice();
+    let time_slice = Duration::from_secs(day_in_secs);
+
+    let airdrop_limiter = AirdropLimiter::new(RwLock::new(neon_token::AirdropLimiter::new(
+        trusted_proxies,
+        blacklist,
+        time_slice,
+        per_request_cap,
+        per_time_cap,
+    )));
+    let airdrop_limiter_1 = airdrop_limiter.clone();
+
+    let airdrop_limiter_reset_stop = Arc::new(std::sync::atomic::AtomicBool::new(false));
+    let airdrop_limiter_reset_stop_1 = airdrop_limiter_reset_stop.clone();
+    let airdrop_limiter_reset = std::thread::spawn(move || {
```

Comment:
What if we use [`tokio::spawn`](https://docs.rs/tokio/latest/tokio/task/fn.spawn.html) so it can reuse the same async runtime?

Reply:
sure can do that, might as well use tokio-timer as well at that point - was going for the stupid and quick solution